clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -56,10 +56,12 @@ public int sizeInBytes() {
 
     @Override
     public long writeTo(GatheringByteChannel channel, long position, int length) throws IOException {
-        ByteBuffer dup = buffer.duplicate();
+        if (position > Integer.MAX_VALUE)
+            throw new IllegalArgumentException("position should not be larger than Integer.MAX_VALUE: " + position);
         int pos = (int) position;
+        ByteBuffer dup = buffer.duplicate();
         dup.position(pos);
-        dup.limit(pos + length);
+        dup.limit(Math.min(pos + length, buffer.limit()));
```

Comment:
It's a fair question. I think it depends on the desired semantics. I changed the code to do the same it did before Jason's changes. The idea is that the duplicate buffer should not extend past the limit of the original buffer (it seems unsafe otherwise). If the underlying buffer is not large enough, `writeTo` will eventually return `0`. This doesn't really help the caller, but at least it doesn't corrupt the underlying buffer.

I think you're right that it's even better to fail-fast if the caller is trying to write more than what this buffer can hold. I'll do that and we'll see what Jason thinks when he reviews the PR.